flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/test/util/MiniClusterExtension.java

```diff
@@ -0,0 +1,340 @@
+ * InjectClusterClient}, {@link InjectClusterRESTAddress}, {@link InjectClusterClientConfiguration}:
+ *
+ * <pre>{@code
+ * class MyTest {
+ *
+ *     @RegisterExtension
+ *     public static final MiniClusterExtension MINI_CLUSTER_RESOURCE =
+ *             new MiniClusterExtension(
+ *                     new MiniClusterResourceConfiguration.Builder()
+ *                             .setNumberTaskManagers(1)
+ *                             .setConfiguration(new Configuration())
+ *                             .build());
+ *
+ *     @Test
+ *     public void myTest(@InjectMiniCluster MiniCluster miniCluster) {
+ *          // Use miniCluster
+ *     }
+ *
+ *     @Test
+ *     public void myTest(@InjectClusterClient ClusterClient<?> clusterClient) {
+ *          // clusterClient here is an instance of MiniClusterClient
+ *     }
+ *
+ *     @Test
+ *     public void myTest(@InjectClusterClient RestClusterClient<?> restClusterClient) {
+ *          // Using RestClusterClient as parameter type will force the creation of a RestClusterClient,
+ *          //  rather than MiniClusterClient
+ *     }
+ * }
+ * }</pre>
+ *
+ * <p>You can use it both with programmatic and declarative extension annotations. Check <a
+ * href="https://junit.org/junit5/docs/snapshot/user-guide/#extensions-registration>JUnit 5 docs</a>
+ * for more details.
+ *
+ * <p>This extension by default creates one {@link MiniCluster} per test class. If you need one
+ * instance of {@link MiniCluster} per test, we suggest you to instantiate it manually in your test
+ * code.
+ *
+ * <p>This extension works correctly with parallel tests and with parametrized tests, but it doesn't
+ * work when combining {@link TestFactory} and parallel tests, because of <a
+ * href="https://github.com/junit-team/junit5/issues/378">some limitations</a>. Use {@link
+ * ParameterizedTest} instead.
+ */
+@Experimental
+public class MiniClusterExtension
+        implements BeforeAllCallback,
+                BeforeEachCallback,
+                AfterEachCallback,
+                AfterAllCallback,
+                ParameterResolver {
+
+    private static final ExtensionContext.Namespace NAMESPACE =
+            ExtensionContext.Namespace.create(MiniClusterExtension.class);
+
+    private static final String CLUSTER_REST_CLIENT = "clusterRestClient";
+    private static final String MINI_CLUSTER_CLIENT = "miniClusterClient";
+
+    /**
+     * Annotate a test method parameter with this annotation to inject the {@link MiniCluster}
+     * instance.
+     */
+    @Target(ElementType.PARAMETER)
+    @Retention(RetentionPolicy.RUNTIME)
+    @Experimental
+    public @interface InjectMiniCluster {}
+
+    /**
+     * Annotate a test method parameter with this annotation to inject the {@link ClusterClient} or
+     * the {@link RestClusterClient} instance.
+     */
+    @Target(ElementType.PARAMETER)
+    @Retention(RetentionPolicy.RUNTIME)
+    @Experimental
+    public @interface InjectClusterClient {}
```

Comment:
Having them as inner classes makes them surprisingly noisy (even more so if we had some in the InternalMiniClusterExtension).

Reply:
Can we have a package for this extension only then? Something like `org.apache.flink.test.junit5`

Reply:
I don't think anything is stopping us from doing that, yes.